File: grammar/js/n3PrintVisitor.js
var n3Visitor = require('./n3Visitor').n3Visitor;

function n3PrintVisitor(output) {
	n3Visitor.call(this);
	this.lvl = 0;
	this.output = output;
	
	return this;
}

n3PrintVisitor.prototype = Object.create(n3Visitor.prototype)
n3PrintVisitor.prototype.constructor = n3PrintVisitor;

n3PrintVisitor.prototype.print = function(el) {	
	var ws = new Array(this.lvl + 1).join(" ");
	var out = ws + el + "\n";
	
	this.output.ast(out);
}

//Visit a parse tree produced by n3Parser#n3Doc.
n3PrintVisitor.prototype.visitN3Doc = function(ctx) {
	this.print("N3Doc");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#n3Statement.
n3PrintVisitor.prototype.visitN3Statement = function(ctx) {
	this.print("N3Statement");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#n3Directive.
n3PrintVisitor.prototype.visitN3Directive = function(ctx) {
	this.print("N3Directive");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#sparqlDirective.
n3PrintVisitor.prototype.visitSparqlDirective = function(ctx) {
	this.print("SparqlDirective");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#sparqlBase.
n3PrintVisitor.prototype.visitSparqlBase = function(ctx) {
	this.print("SparqlBase");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#sparqlPrefix.
n3PrintVisitor.prototype.visitSparqlPrefix = function(ctx) {
	this.print("SparqlPrefix");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#prefixID.
n3PrintVisitor.prototype.visitPrefixID = function(ctx) {
	this.print("PrefixID");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#base.
n3PrintVisitor.prototype.visitBase = function(ctx) {
	this.print("Base");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#triples.
n3PrintVisitor.prototype.visitTriples = function(ctx) {
	this.print("Triples");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#predicateObjectList.
n3PrintVisitor.prototype.visitPredicateObjectList = function(ctx) {
	this.print("PredicateObjectList");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#objectList.
n3PrintVisitor.prototype.visitObjectList = function(ctx) {
	this.print("ObjectList");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#verb.
n3PrintVisitor.prototype.visitVerb = function(ctx) {
	this.print("Verb");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#subject.
n3PrintVisitor.prototype.visitSubject = function(ctx) {
	this.print("Subject");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#predicate.
n3PrintVisitor.prototype.visitPredicate = function(ctx) {
	this.print("Predicate");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#object.
n3PrintVisitor.prototype.visitObject = function(ctx) {
	this.print("Object");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#expression.
n3PrintVisitor.prototype.visitExpression = function(ctx) {
	this.print("Expression");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#path.
n3PrintVisitor.prototype.visitPath = function(ctx) {
	this.print("Path");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#pathItem.
n3PrintVisitor.prototype.visitPathItem = function(ctx) {
	this.print("PathItem");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#literal.
n3PrintVisitor.prototype.visitLiteral = function(ctx) {
	this.print("Literal");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#blankNodePropertyList.
n3PrintVisitor.prototype.visitBlankNodePropertyList = function(ctx) {
	this.print("BlankNodePropertyList");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#collection.
n3PrintVisitor.prototype.visitCollection = function(ctx) {
	this.print("Collection");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#formula.
n3PrintVisitor.prototype.visitFormula = function(ctx) {
	this.print("Formula");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#formulaContent.
n3PrintVisitor.prototype.visitFormulaContent = function(ctx) {
	this.print("FormulaContent");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#rdfLiteral.
n3PrintVisitor.prototype.visitRdfLiteral = function(ctx) {
	this.print("RdfLiteral");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#iri.
n3PrintVisitor.prototype.visitIri = function(ctx) {
	this.print("Iri");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#iriList.
n3PrintVisitor.prototype.visitIriList = function(ctx) {
	this.print("IriList");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#prefixedName.
n3PrintVisitor.prototype.visitPrefixedName = function(ctx) {
	this.print("PrefixedName");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#existential.
n3PrintVisitor.prototype.visitExistential = function(ctx) {
	this.print("Existential");
	return this.doVisitChildren(ctx);
};

// Visit a parse tree produced by n3Parser#universal.
n3PrintVisitor.prototype.visitUniversal = function(ctx) {
	this.print("Universal");
	return this.doVisitChildren(ctx);
};

n3PrintVisitor.prototype.doVisitChildren = function(ctx) {
	this.lvl++;
	var ret = this.visitChildren(ctx);
	this.lvl--;
	
	return ret;
}

exports.n3PrintVisitor = n3PrintVisitor;

File: grammar/src/main/java/w3c/n3dev/parser/visitor/n3/n3PrintVisitor.java
package w3c.n3dev.parser.visitor.n3;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.RuleNode;
import org.antlr.v4.runtime.tree.TerminalNode;

import w3c.n3dev.parser.antlr.n3Parser;
import wvw.utils.log.Log;

public class n3PrintVisitor extends n3DefaultVisitor {

	private int lvl = 0;

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitN3Doc(n3Parser.N3DocContext ctx) {
		print("N3Doc");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitN3Statement(n3Parser.N3StatementContext ctx) {
		print("N3Statement");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitN3Directive(n3Parser.N3DirectiveContext ctx) {
		print("N3Directive");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitSparqlDirective(n3Parser.SparqlDirectiveContext ctx) {
		print("SparqlDirective");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitSparqlBase(n3Parser.SparqlBaseContext ctx) {
		print("SparqlBase");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitSparqlPrefix(n3Parser.SparqlPrefixContext ctx) {
		print("SparqlPrefix");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitPrefixID(n3Parser.PrefixIDContext ctx) {
		print("PrefixID");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitBase(n3Parser.BaseContext ctx) {
		print("Base");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitTriples(n3Parser.TriplesContext ctx) {
		print("Triples");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitPredicateObjectList(n3Parser.PredicateObjectListContext ctx) {
		print("PredicateObjectList");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitObjectList(n3Parser.ObjectListContext ctx) {
		print("ObjectList");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitVerb(n3Parser.VerbContext ctx) {
		print("Verb");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitSubject(n3Parser.SubjectContext ctx) {
		print("Subject");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitPredicate(n3Parser.PredicateContext ctx) {
		print("Predicate");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitObject(n3Parser.ObjectContext ctx) {
		print("Object");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitExpression(n3Parser.ExpressionContext ctx) {
		print("Expression");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitPath(n3Parser.PathContext ctx) {
		print("Path");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitPathItem(n3Parser.PathItemContext ctx) {
		print("PathItem");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitLiteral(n3Parser.LiteralContext ctx) {
		print("Literal");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitBlankNodePropertyList(n3Parser.BlankNodePropertyListContext ctx) {
		print("BlankNodePropertyList");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitCollection(n3Parser.CollectionContext ctx) {
		print("Collection");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitFormula(n3Parser.FormulaContext ctx) {
		print("Formula");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitFormulaContent(n3Parser.FormulaContentContext ctx) {
		print("FormulaContent");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitRdfLiteral(n3Parser.RdfLiteralContext ctx) {
		print("RdfLiteral");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitIri(n3Parser.IriContext ctx) {
		print("Iri");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitIriList(n3Parser.IriListContext ctx) {
		print("IriList");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitPrefixedName(n3Parser.PrefixedNameContext ctx) {
		print("PrefixedName");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitExistential(n3Parser.ExistentialContext ctx) {
		print("Existential");
		return doVisitChildren(ctx);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * The default implementation returns the result of calling
	 * {@link #visitChildren} on {@code ctx}.
	 * </p>
	 */
	@Override
	public Void visitUniversal(n3Parser.UniversalContext ctx) {
		print("Universal");
		return doVisitChildren(ctx);
	}

	private void print(String element) {
		StringBuffer str = new StringBuffer();
		for (int i = 0; i < lvl; i++)
			str.append(" ");

		str.append(element);
		Log.i(str);
	}

	private Void doVisitChildren(RuleNode ctx) {
		lvl++;
		Void ret = visitChildren(ctx);
		lvl--;

		return ret;
	}

	@Override
	public Void visitChildren(RuleNode node) {
		Void result = defaultResult();
		int n = node.getChildCount();
		for (int i = 0; i < n; i++) {
			if (!shouldVisitNextChild(node, result)) {
				break;
			}

			ParseTree c = node.getChild(i);
			if (c instanceof TerminalNode) {
				TerminalNode t = (TerminalNode) c;
				String out = c.toString();
				int type = t.getSymbol().getType();
				if (type != -1)
					out += " (" + n3Parser.tokenNames[type] + ")";
				print(out);

			} else {
				Void childResult = c.accept(this);
				result = aggregateResult(result, childResult);
			}
		}

		return result;
	}
}
